lib/xwax/pitch_kalman.h

```diff
@@ -0,0 +1,325 @@
+
+    p->x = 0.0;
+    p->v = 0.0;
+
+    /* Start with large uncertainty so the filter trusts early measurements */
+
+    p->P[x][x] = 1e6;
+    p->P[x][v] = 0.0;
+    p->P[v][v] = 1e6;
+
+    /* Fixed thresholds for the mode switches */
+
+    p->scratch_threshold = scratch_threshold;
+    p->medium_threshold = medium_threshold;
+
+    /* Q and R for the different modes */
+
+    p->stable = stable;
+    p->medium = medium;
+    p->reactive = reactive;
+
+    /* Initialize as reactive */
+
+    kalman_tune_sensitivity(p, &p->reactive);
+}
+
+/*
+ * Feed one observation: in the last dt seconds, position moved by dx
+ */
+
+static inline void pitch_kalman_update(struct pitch_kalman* p, double dx) {
+
+    if (!p) {
+        errno = EINVAL;
+        perror(__func__);
+        return;
+    }
+
+    /* Sampling interval */
+
+    const double dt = p->dt;
+
+    /* Indexes for the matrices */
+
+    const size_t x = 0;
+    const size_t v = 1;
+
+    /*
+     * =================== Prediction step =====================
+     *
+     * State transition for constant-velocity model:
+     *   F = [ 1  dt
+     *         0  1 ]
+     *
+     * Discretizing, the process noise covariance is:
+     *   Q = Q * [ dt^3/3  dt^2/2
+     *             dt^2/2  dt     ]
+     *
+     * Here p->coeffs->Q is 'Q', the acceleration noise variance.
+     * Q[v][x] is not needed for the computations, therefore we put 0.0.
+     */
+
+    const double F[2][2] = { { 1.0,   dt },
+                             { 0.0,  1.0 }};
+
+    const double Q[2][2] = { { p->coeffs->Q * (pow(dt, 3) / 3.0),  p->coeffs->Q * (pow(dt, 2) / 2.0) },
+                             {                               0.0,  p->coeffs->Q * dt                 }};
+
+    /* Predict state */
+
+    const double x_pred = p->x + p->v * dt;
+    const double v_pred = p->v;
+
+    /*
+     * Predict covariance: P_pred = F * P * F^T + Q
+     * For 2×2 P and F, expand each term explicitly:
+     *
+     *   P_pred[x][x] = F00 * (F00 * P[x][x] + F01 * P[x][v]) + F01 * (F00 * P[x][v] + F01 * P[v][v]) + Q11
+     *   P_pred[x][v] = F10 * (F00 * P[x][x] + F01 * P[x][v]) + F11 * (F00 * P[x][v] + F01 * P[v][v]) + Q12
+     *   P_pred[v][v] = F10 * (F10 * P[x][x] + F11 * P[x][v]) + F11 * (F10 * P[x][v] + F11 * P[v][v]) + Q22
+     *
+     * This preserves symmetry while avoiding full matrix multiplication.
+     */
+
+    const double P[2][2] = { {p->P[x][x], p->P[x][v]},
+                             {p->P[v][x], p->P[v][v]} };
+
+    double P_pred[2][2];
+
+    P_pred[x][x] = F[x][x] * (F[x][x] * P[x][x] + F[x][v] * P[x][v]) +
+                   F[x][v] * (F[x][x] * P[x][v] + F[x][v] * P[v][v]) + Q[x][x];
+
+    P_pred[x][v] = F[v][x] * (F[x][x] * P[x][x] + F[x][v] * P[x][v]) +
+                   F[v][v] * (F[x][x] * P[x][v] + F[x][v] * P[v][v]) + Q[x][v];
+
+    P_pred[v][v] = F[v][x] * (F[v][x] * P[x][x] + F[v][v] * P[x][v]) +
+                   F[v][v] * (F[v][x] * P[x][v] + F[v][v] * P[v][v]) + Q[v][v];
+
```

Comment:
If you're modelling the entire matrix, then also initialize the symmetric element.


```suggestion
    P_pred[v][v] = F[v][x] * (F[v][x] * P[x][x] + F[v][v] * P[x][v]) +
                   F[v][v] * (F[v][x] * P[x][v] + F[v][v] * P[v][v]) + Q[v][v];
    P_pred[v][x] = P_pred[x][v];
```

Reply:
It's a draft